programs/stake/src/stake_state.rs

```diff
@@ -542,8 +544,21 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
         match self.state()? {
             StakeState::Initialized(meta) => {
                 meta.authorized.check(signers, StakeAuthorize::Staker)?;
+                let stake_amount = self.lamports()?.saturating_sub(meta.rent_exempt_reserve); // can't stake the rent ;)
+                if stake_amount < MINIMUM_STAKE_DELEGATION {
+                    // Since (1) delegation uses the full balance (minus rent_exempt_reserve), and
+                    // (2) stake accounts must be created with a minimum balance of
+                    // rent_exempt_reserve + MINIMUM_STAKE_DELEGATION, then we can be sure that the
+                    // stake_amount will always be at least MINIMUM_STAKE_DELEGATION.
+                    //
+                    // This error case can only be triggered if (1) the MINIMUM_STAKE_DELEGATION is
+                    // raised in the future and this account was created before the raise, or (2)
+                    // during tests where direct manipulation of the account allows us to break
+                    // invariants.
+                    return Err(InstructionError::StakeDelegationTooSmall);
+                }
```

Comment:
how about factoring out this into `fn validate_stake_amount()` just like `fn valiate_split_amount()` to avoid this somewhat long comment copy-paste here?: https://github.com/solana-labs/solana/pull/22663/files?diff=split&w=1#r802606743